# Content List

## 1. Import Libraries

## 2. Dataset: Frequency Internet Use
### 2.1 Import Data
### 2.2 Data Consistency Check
### 2.3 Dropping Columns
### 2.4 Renaming Columns
### 2.5 Removing Prefixes
### 2.6 Mixed Datatypes
### 2.7 Missing Data
### 2.8 Duplicated Values
### 2.9 Exporting Data

## 3. Dataset: Internet Activities
### 3.1 Import Data
### 3.2 Data Consistency Check
### 3.3 Dropping Columns
### 3.4 Renaming Columns
### 3.5 Removing Prefixes
### 3.6 Mixed Datatypes
### 3.7 Missing Data
#### 3.7.1 Dataset internet activitites
#### 3.7.2 Dataset internet activities 2
#### 3.7.3 Dataset internet activities 3
### 3.8 Duplicated Values
### 3.9 Exporting Data

## 4. Dataset: Purchasing Power GDP
### 4.1 Import Data
### 4.2 Data Consistency Check
### 4.3 Dropping Columns
### 4.4 Renaming Columns
### 4.5 Mixed Datatypes
### 4.6 Missing Data
### 4.7 Duplicated Values
### 4.8 Exporting Data

# 01. Import Libraries

In [1]:
# Import libraries 
import pandas as pd
import numpy as np
import os

# 02. Dataset: Frequency Internet Use

## 02.1 Import Data

In [2]:
# Create path to use as shortcut to access project folder for importing data
path = r'C:\Users\adrig\OneDrive\Documentos\CAREERFOUNDRY\DATA IMMERSION\ACHIEVEMENT 6\Internet use in EU'
path

'C:\\Users\\adrig\\OneDrive\\Documentos\\CAREERFOUNDRY\\DATA IMMERSION\\ACHIEVEMENT 6\\Internet use in EU'

In [3]:
# Import main dataset (frequency of internet use) 
internet_use = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'frequency_internet_use.csv'), index_col = False)

C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\327273434.py:2: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  internet_use = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'frequency_internet_use.csv'), index_col = False)


## 02.2 Data Consistency Check

In [4]:
# Dimensions of dataset
internet_use.shape

(92149, 11)

In [5]:
# Column names
internet_use.columns

Index(['DATAFLOW', 'LAST UPDATE', 'freq', 'indic_is', 'unit', 'ind_type',
       'geo', 'TIME_PERIOD', 'OBS_VALUE', 'OBS_FLAG', 'CONF_STATUS'],
      dtype='object')

In [6]:
# first rows 
internet_use.head()

DATAFLOW        LAST UPDATE    freq  \
0  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
1  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
2  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
3  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
4  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   

                              indic_is                       unit  \
0  Frequency of internet access: daily  Percentage of individuals   
1  Frequency of internet access: daily  Percentage of individuals   
2  Frequency of internet access: daily  Percentage of individuals   
3  Frequency of internet access: daily  Percentage of individuals   
4  Frequency of internet access: daily  Percentage of individuals   

                                            ind_type      geo  TIME_PERIOD  \
0  Persons aged 16-74 working in agriculture, for...  Austria         2021   
1  Persons aged 16-74 working in agriculture, for...  Austria         2022   
2  Persons aged 16-74 working in agriculture, for...  Austria         2023   
3  Persons aged 16-74 working in agriculture, for...  Austria         2024   
4  Persons aged 16-74 working in agriculture, for...  Belgium         2021   

   OBS_VALUE OBS_FLAG CONF_STATUS  
0      71.67      NaN         NaN  
1      74.32      NaN         NaN  
2      78.30      NaN         NaN  
3      65.60      NaN         NaN  
4      73.35        u         NaN

There are unnecessary columns that need to be deleted. 

In [7]:
# General information about dataser
internet_use.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 92149 entries, 0 to 92148
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   DATAFLOW     92149 non-null  object 
 1   LAST UPDATE  92149 non-null  object 
 2   freq         92149 non-null  object 
 3   indic_is     92149 non-null  object 
 4   unit         92149 non-null  object 
 5   ind_type     92149 non-null  object 
 6   geo          92149 non-null  object 
 7   TIME_PERIOD  92149 non-null  int64  
 8   OBS_VALUE    88519 non-null  float64
 9   OBS_FLAG     6057 non-null   object 
 10  CONF_STATUS  9 non-null      object 
dtypes: float64(1), int64(1), object(9)
memory usage: 7.7+ MB


6057 rows have an observation flag. 
The variable CONF_STATUS has 9 non-null values, therefore should be checked. 

In [8]:
# descriptive stats 
internet_use.describe()

TIME_PERIOD     OBS_VALUE
count  92149.000000  88519.000000
mean    2014.348023     24.680962
std        6.053127     33.815669
min     2003.000000      0.000000
25%     2009.000000      1.640000
50%     2014.000000      5.580000
75%     2020.000000     42.645000
max     2024.000000    100.000000

With the 'count' values, missing values are suggested in the OBS_VALUE column. 
The years shown are from 2003 to 2024
Values (percentages) are correct, as the minimum is 0 and the max 100

## 02.3 Dropping Columns

In [9]:
internet_use['DATAFLOW'].value_counts(dropna = False)

DATAFLOW
ESTAT:ISOC_CI_IFP_FU(1.0)    92149
Name: count, dtype: int64

In [10]:
internet_use['LAST UPDATE'].value_counts(dropna = False)

LAST UPDATE
24/06/25 11:00:00    92149
Name: count, dtype: int64

In [11]:
internet_use['freq'].value_counts(dropna = False)

freq
Annual    92149
Name: count, dtype: int64

In [12]:
internet_use['unit'].value_counts(dropna = False)

unit
Percentage of individuals    92149
Name: count, dtype: int64

In [13]:
internet_use['CONF_STATUS'].value_counts(dropna = False)

CONF_STATUS
NaN    92140
C          9
Name: count, dtype: int64

In [14]:
df_c = internet_use[internet_use['CONF_STATUS'] == 'C']
df_c

DATAFLOW        LAST UPDATE    freq  \
28300  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
31052  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
32335  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
32827  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
46193  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
46194  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
62929  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
62930  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   
62931  ESTAT:ISOC_CI_IFP_FU(1.0)  24/06/25 11:00:00  Annual   

                                                indic_is  \
28300  Frequency of internet access: less than once a...   
31052  Frequency of internet access: less than once a...   
32335  Frequency of internet access: less than once a...   
32827  Frequency of internet access: less than once a...   
46193  Frequency of internet access: less than once a...   
46194  Frequency of internet access: less than once a...   
62929  Frequency of internet access: at least once a ...   
62930  Frequency of internet access: at least once a ...   
62931  Frequency of internet access: at least once a ...   

                            unit                     ind_type      geo  \
28300  Percentage of individuals  Females, 16 to 24 years old  Germany   
31052  Percentage of individuals            ICT professionals  Germany   
32335  Percentage of individuals    Males, 16 to 24 years old  Germany   
32827  Percentage of individuals                     Students  Germany   
46193  Percentage of individuals            ICT professionals  Germany   
46194  Percentage of individuals            ICT professionals  Germany   
62929  Percentage of individuals            ICT professionals  Germany   
62930  Percentage of individuals            ICT professionals  Germany   
62931  Percentage of individuals            ICT professionals  Germany   

       TIME_PERIOD  OBS_VALUE OBS_FLAG CONF_STATUS  
28300         2009        NaN      NaN           C  
31052         2009        NaN      NaN           C  
32335         2010        NaN      NaN           C  
32827         2010        NaN      NaN           C  
46193         2007        NaN      NaN           C  
46194         2008        NaN      NaN           C  
62929         2007        NaN      NaN           C  
62930         2008        NaN      NaN           C  
62931         2009        NaN      NaN           C

The column will be eliminated together with the entire rows, as it provides no relevant information.

In [15]:
# Delete rows with CONF_STATUS value: 'C'
internet_use = internet_use[internet_use['CONF_STATUS'] != 'C']

In [16]:
# Checking deletion
internet_use['CONF_STATUS'].value_counts(dropna=False)

CONF_STATUS
NaN    92140
Name: count, dtype: int64

In [17]:
internet_use['OBS_FLAG'].value_counts(dropna = False)

OBS_FLAG
NaN    86083
u       4744
b       1260
bu        53
Name: count, dtype: int64

This column would not be dropped, as it contains information about the values presented on those rows (flag letters). 

In [18]:
# Drop unnecessary columns: DATAFLOW / LAST UPDATE / freq / unit / CONF_STATUS
internet_use = internet_use.drop(columns = ['DATAFLOW', 'LAST UPDATE', 'freq', 'unit', 'CONF_STATUS'])

In [19]:
internet_use.head()

indic_is  \
0  Frequency of internet access: daily   
1  Frequency of internet access: daily   
2  Frequency of internet access: daily   
3  Frequency of internet access: daily   
4  Frequency of internet access: daily   

                                            ind_type      geo  TIME_PERIOD  \
0  Persons aged 16-74 working in agriculture, for...  Austria         2021   
1  Persons aged 16-74 working in agriculture, for...  Austria         2022   
2  Persons aged 16-74 working in agriculture, for...  Austria         2023   
3  Persons aged 16-74 working in agriculture, for...  Austria         2024   
4  Persons aged 16-74 working in agriculture, for...  Belgium         2021   

   OBS_VALUE OBS_FLAG  
0      71.67      NaN  
1      74.32      NaN  
2      78.30      NaN  
3      65.60      NaN  
4      73.35        u

## 02.4 Renaming Columns

In [20]:
# Renaming columns for better understanding and same format
internet_use.rename(columns = {
    'indic_is' : 'freq_internet_access',
    'ind_type' : 'demographic_group',
    'geo' : 'country',
    'TIME_PERIOD' : 'year',
    'OBS_VALUE' : 'internet_access_rate',
    'OBS_FLAG' : 'freq_data_flag'
}, inplace = True)

In [21]:
internet_use.head()

freq_internet_access  \
0  Frequency of internet access: daily   
1  Frequency of internet access: daily   
2  Frequency of internet access: daily   
3  Frequency of internet access: daily   
4  Frequency of internet access: daily   

                                   demographic_group  country  year  \
0  Persons aged 16-74 working in agriculture, for...  Austria  2021   
1  Persons aged 16-74 working in agriculture, for...  Austria  2022   
2  Persons aged 16-74 working in agriculture, for...  Austria  2023   
3  Persons aged 16-74 working in agriculture, for...  Austria  2024   
4  Persons aged 16-74 working in agriculture, for...  Belgium  2021   

   internet_access_rate freq_data_flag  
0                 71.67            NaN  
1                 74.32            NaN  
2                 78.30            NaN  
3                 65.60            NaN  
4                 73.35              u

## 02.5 Removing prefixes

In [22]:
# Checking if all values in the variable 'freq_internet_access' have the same prefix - Frequency of internet access -
internet_use['freq_internet_access'].value_counts(dropna = False)

freq_internet_access
Frequency of internet access: daily                                       27430
Frequency of internet access: at least once a week (but not every day)    24736
Frequency of internet access: less than once a week                       23779
Frequency of internet access: at least once a month                        8117
Frequency of internet access: less than once a month                       8078
Name: count, dtype: int64

In [23]:
# Removing prefix 
internet_use['freq_internet_access'] = internet_use['freq_internet_access'].str.replace('Frequency of internet access: ', '')

In [24]:
internet_use['freq_internet_access'].value_counts(dropna = False)

freq_internet_access
daily                                       27430
at least once a week (but not every day)    24736
less than once a week                       23779
at least once a month                        8117
less than once a month                       8078
Name: count, dtype: int64

## 02.6 Mixed Data Types

In [25]:
# function to check for mixed-type columns
for col in internet_use.columns.tolist():
    weird = (internet_use[[col]].applymap(type) != internet_use[[col]].iloc[0].apply(type)).any(axis=1)
    if len (internet_use[weird]) > 0:
        print(col)

freq_data_flag


C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\4204843813.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_use[[col]].applymap(type) != internet_use[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\4204843813.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_use[[col]].applymap(type) != internet_use[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\4204843813.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_use[[col]].applymap(type) != internet_use[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\4204843813.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_use[[col]].applymap(type) != internet_use[[col]].iloc[0].apply(type)).any(axis

In [26]:
# Changing data type for freq_data_flag variable
internet_use['freq_data_flag'] = internet_use['freq_data_flag'].astype('category')

In [27]:
internet_use.dtypes

freq_internet_access      object
demographic_group         object
country                   object
year                       int64
internet_access_rate     float64
freq_data_flag          category
dtype: object

## 02.7 Missing Values

In [28]:
# Finding missing values in internet_use dataframe
internet_use.isnull().sum()

freq_internet_access        0
demographic_group           0
country                     0
year                        0
internet_access_rate     3621
freq_data_flag          86083
dtype: int64

In [29]:
# Creating subset for NaN values 
df_nan = internet_use[internet_use['internet_access_rate'].isnull() == True]
df_nan

freq_internet_access  \
27                                        daily   
55                                        daily   
57                                        daily   
83                                        daily   
91                                        daily   
...                                         ...   
92141  at least once a week (but not every day)   
92142  at least once a week (but not every day)   
92143  at least once a week (but not every day)   
92144  at least once a week (but not every day)   
92145  at least once a week (but not every day)   

                                       demographic_group         country  \
27     Persons aged 16-74 working in agriculture, for...         Denmark   
55     Persons aged 16-74 working in agriculture, for...      Luxembourg   
57     Persons aged 16-74 working in agriculture, for...      Luxembourg   
83     Persons aged 16-74 working in agriculture, for...        Slovenia   
91     Persons aged 16-74 working in mining or quarrying         Austria   
...                                                  ...             ...   
92141                  Individuals, 75 years old or more  United Kingdom   
92142                  Individuals, 75 years old or more  United Kingdom   
92143                  Individuals, 75 years old or more  United Kingdom   
92144                  Individuals, 75 years old or more  United Kingdom   
92145                  Individuals, 75 years old or more  United Kingdom   

       year  internet_access_rate freq_data_flag  
27     2024                   NaN              u  
55     2021                   NaN              u  
57     2023                   NaN              u  
83     2021                   NaN              u  
91     2021                   NaN              u  
...     ...                   ...            ...  
92141  2005                   NaN              u  
92142  2006                   NaN              u  
92143  2007                   NaN              u  
92144  2008                   NaN              u  
92145  2009                   NaN              u  

[3621 rows x 6 columns]

In [30]:
# Checking missing values in internet_access_rate
internet_use[internet_use['internet_access_rate'].isnull()]['country'].value_counts()

country
United Kingdom     391
Portugal           346
North Macedonia    308
Germany            301
Iceland            293
Austria            243
Estonia            206
Italy              196
Finland            174
Slovenia           157
Bulgaria           127
Czechia            118
Ireland            104
Denmark             98
Romania             88
Poland              83
Malta               56
Norway              47
Lithuania           36
Luxembourg          32
Greece              29
Sweden              28
Slovakia            28
Hungary             23
Cyprus              19
Croatia             16
Montenegro          12
Albania             12
Netherlands         12
Belgium             10
Latvia               9
France               9
Serbia               6
Kosovo*              3
Spain                1
Name: count, dtype: int64

In [31]:
# Checking missing values in internet_access_rate
internet_use[internet_use['internet_access_rate'].isnull()]['demographic_group'].value_counts()

demographic_group
ICT professionals                                                                                              410
Individuals, 65 to 74 years old                                                                                192
Unemployed                                                                                                     179
Males with high formal education                                                                               144
Students                                                                                                       143
Females, 16 to 24 years old                                                                                    141
Males, 16 to 24 years old                                                                                      139
Individuals, 75 years old or more                                                                              130
Males with low formal education                               

In [32]:
# Checking missing values in internet_access_rate
internet_use[internet_use['internet_access_rate'].isnull()]['year'].value_counts()

year
2009    563
2008    510
2010    410
2005    408
2007    407
2006    384
2004    194
2021     91
2022     79
2024     74
2023     74
2003     68
2015     63
2012     45
2011     45
2018     44
2017     43
2019     34
2020     31
2013     21
2016     21
2014     12
Name: count, dtype: int64

In [33]:
# Checking missing values in internet_access_rate
internet_use[internet_use['internet_access_rate'].isnull()].groupby(['country', 'demographic_group', 'year']).size().sort_values(ascending=False)

country      demographic_group                                  year
Luxembourg   ICT professionals                                  2011    5
Estonia      Individuals living in towns and suburbs            2005    5
Italy        ICT professionals                                  2007    5
                                                                2008    5
                                                                2009    5
                                                                       ..
Portugal     Individuals with medium formal education           2004    1
Netherlands  Persons aged 16-74 working in mining or quarrying  2023    1
                                                                2022    1
Portugal     Individuals with medium formal education           2008    1
Czechia      Females with low formal education                  2007    1
Length: 1639, dtype: int64

Due to the fact that the missing values are 3.9% of the total data, and it is distributed across countries, demographic groups and years, I decided to delete the rows with missing data.

In [34]:
# Removing rows
internet_use_clean = internet_use[internet_use['internet_access_rate'].isnull() == False]

In [35]:
internet_use_clean.shape

(88519, 6)

In [36]:
# Finding missing values in internet_use dataframe
internet_use_clean.isnull().sum()

freq_internet_access        0
demographic_group           0
country                     0
year                        0
internet_access_rate        0
freq_data_flag          86083
dtype: int64

## 02.8 Duplicate values

In [37]:
# Creating a new subset 'df_dups' containing just duplicated values
df_dups = internet_use_clean[internet_use_clean.duplicated()]

In [38]:
df_dups

Empty DataFrame
Columns: [freq_internet_access, demographic_group, country, year, internet_access_rate, freq_data_flag]
Index: []

There are no duplicate values in the dataset. 

## 02.9 Exporting Data

In [39]:
# Exporting first dataframe internet_use_clean
internet_use_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'freq_internet_use_checked.csv'))

# 03. Dataset: Internet Activities

This data is divided in three different datasets, because the webpage only allow to download certain number of rows and the information from 2003 to 2024 exceded it. So it is divided in three datasets 2003-2009, 2010-2019, 2020-2024. 

## 03.1 Import Data

In [40]:
# Import dataset (internet activities) 
internet_activities = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'internet_activities_2003-09.csv'), index_col = False)

In [41]:
# Import dataset (internet activities) 
internet_activities2 = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'internet_activities_2010-19.csv'), index_col = False)

C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\2494412370.py:2: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  internet_activities2 = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'internet_activities_2010-19.csv'), index_col = False)


In [42]:
# Import dataset (internet activities) 
internet_activities3 = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'internet_activities_2020-24.csv'), index_col = False)

C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\3304253281.py:2: DtypeWarning: Columns (10) have mixed types. Specify dtype option on import or set low_memory=False.
  internet_activities3 = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'internet_activities_2020-24.csv'), index_col = False)


## 03.2 Data Consistency Check

In [43]:
# dimensions of dataframe
internet_activities.shape

(56474, 11)

In [44]:
# dimensions of dataframe
internet_activities2.shape

(153420, 11)

In [45]:
# dimensions of dataframe
internet_activities3.shape

(102937, 11)

In [46]:
# names of variables
internet_activities.columns

Index(['DATAFLOW', 'LAST UPDATE', 'freq', 'indic_is', 'unit', 'ind_type',
       'geo', 'TIME_PERIOD', 'OBS_VALUE', 'OBS_FLAG', 'CONF_STATUS'],
      dtype='object')

In [47]:
# names of variables
internet_activities2.columns

Index(['DATAFLOW', 'LAST UPDATE', 'freq', 'indic_is', 'unit', 'ind_type',
       'geo', 'TIME_PERIOD', 'OBS_VALUE', 'OBS_FLAG', 'CONF_STATUS'],
      dtype='object')

In [48]:
# names of variables
internet_activities3.columns

Index(['DATAFLOW', 'LAST UPDATE', 'freq', 'indic_is', 'unit', 'ind_type',
       'geo', 'TIME_PERIOD', 'OBS_VALUE', 'OBS_FLAG', 'CONF_STATUS'],
      dtype='object')

The three datasets contain the same column names. 
There are unnecessary columns that should be deleted. 
A format standarization of the names of columns is needed. 

In [49]:
# first rows
internet_activities.head()

DATAFLOW        LAST UPDATE    freq  \
0  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
1  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
2  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
3  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
4  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   

                         indic_is                       unit  \
0  Internet use: Internet banking  Percentage of individuals   
1  Internet use: Internet banking  Percentage of individuals   
2  Internet use: Internet banking  Percentage of individuals   
3  Internet use: Internet banking  Percentage of individuals   
4  Internet use: Internet banking  Percentage of individuals   

                            ind_type      geo  TIME_PERIOD  OBS_VALUE  \
0  Females with low formal education  Austria         2004       3.94   
1  Females with low formal education  Austria         2005       4.71   
2  Females with low formal education  Austria         2006       5.29   
3  Females with low formal education  Austria         2007        NaN   
4  Females with low formal education  Austria         2008       7.94   

  OBS_FLAG CONF_STATUS  
0      NaN         NaN  
1      NaN         NaN  
2      NaN         NaN  
3        u         NaN  
4      NaN         NaN

In [50]:
# first rows
internet_activities2.head()

DATAFLOW        LAST UPDATE    freq  \
0  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
1  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
2  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
3  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
4  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   

                         indic_is                       unit  \
0  Internet use: Internet banking  Percentage of individuals   
1  Internet use: Internet banking  Percentage of individuals   
2  Internet use: Internet banking  Percentage of individuals   
3  Internet use: Internet banking  Percentage of individuals   
4  Internet use: Internet banking  Percentage of individuals   

                      ind_type      geo  TIME_PERIOD  OBS_VALUE OBS_FLAG  \
0  Employees working full time  Albania         2018       2.83      NaN   
1  Employees working full time  Albania         2019       2.72      NaN   
2  Employees working full time  Austria         2017      69.41      NaN   
3  Employees working full time  Austria         2018      69.62      NaN   
4  Employees working full time  Austria         2019      73.70      NaN   

  CONF_STATUS  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN

In [51]:
# first rows
internet_activities3.head()

DATAFLOW        LAST UPDATE    freq  \
0  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
1  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
2  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
3  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
4  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   

                                            indic_is  \
0  Internet use: accessing personal health record...   
1  Internet use: accessing personal health record...   
2  Internet use: accessing personal health record...   
3  Internet use: accessing personal health record...   
4  Internet use: accessing personal health record...   

                        unit  \
0  Percentage of individuals   
1  Percentage of individuals   
2  Percentage of individuals   
3  Percentage of individuals   
4  Percentage of individuals   

                                            ind_type       geo  TIME_PERIOD  \
0  Persons aged 16-74 working in agriculture, for...   Austria         2022   
1  Persons aged 16-74 working in agriculture, for...   Austria         2024   
2  Persons aged 16-74 working in agriculture, for...   Belgium         2022   
3  Persons aged 16-74 working in agriculture, for...   Belgium         2024   
4  Persons aged 16-74 working in agriculture, for...  Bulgaria         2022   

   OBS_VALUE OBS_FLAG CONF_STATUS  
0      12.47      NaN         NaN  
1       9.03      NaN         NaN  
2      14.77        u         NaN  
3      46.79        u         NaN  
4      10.92      NaN         NaN

In [52]:
# general information 
internet_activities.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 56474 entries, 0 to 56473
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   DATAFLOW     56474 non-null  object 
 1   LAST UPDATE  56474 non-null  object 
 2   freq         56474 non-null  object 
 3   indic_is     56474 non-null  object 
 4   unit         56474 non-null  object 
 5   ind_type     56474 non-null  object 
 6   geo          56474 non-null  object 
 7   TIME_PERIOD  56474 non-null  int64  
 8   OBS_VALUE    53377 non-null  float64
 9   OBS_FLAG     3145 non-null   object 
 10  CONF_STATUS  3 non-null      object 
dtypes: float64(1), int64(1), object(9)
memory usage: 4.7+ MB


3145 rows in the dataframe have an observation flag. 
The variable CONF_STATUS has 3 non-null value that should be checked. 

In [54]:
# general information 
internet_activities2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 153420 entries, 0 to 153419
Data columns (total 11 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   DATAFLOW     153420 non-null  object 
 1   LAST UPDATE  153420 non-null  object 
 2   freq         153420 non-null  object 
 3   indic_is     153420 non-null  object 
 4   unit         153420 non-null  object 
 5   ind_type     153420 non-null  object 
 6   geo          153420 non-null  object 
 7   TIME_PERIOD  153420 non-null  int64  
 8   OBS_VALUE    151587 non-null  float64
 9   OBS_FLAG     7696 non-null    object 
 10  CONF_STATUS  1 non-null       object 
dtypes: float64(1), int64(1), object(9)
memory usage: 12.9+ MB


7696 rows in the dataframe have an observation flag. 
The variable CONF_STATUS has 1 non-null value that should be checked. 

In [55]:
# general information 
internet_activities3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 102937 entries, 0 to 102936
Data columns (total 11 columns):
 #   Column       Non-Null Count   Dtype  
---  ------       --------------   -----  
 0   DATAFLOW     102937 non-null  object 
 1   LAST UPDATE  102937 non-null  object 
 2   freq         102937 non-null  object 
 3   indic_is     102937 non-null  object 
 4   unit         102937 non-null  object 
 5   ind_type     102937 non-null  object 
 6   geo          102937 non-null  object 
 7   TIME_PERIOD  102937 non-null  int64  
 8   OBS_VALUE    100156 non-null  float64
 9   OBS_FLAG     10384 non-null   object 
 10  CONF_STATUS  45 non-null      object 
dtypes: float64(1), int64(1), object(9)
memory usage: 8.6+ MB


10384 rows in the dataframe have an observation flag. 
The variable CONF_STATUS has 45 non-null value that should be checked. 

In [56]:
# descriptive stats
internet_activities.describe()

TIME_PERIOD     OBS_VALUE
count  56474.000000  53377.000000
mean    2006.897386     27.154227
std        1.787066     23.484212
min     2003.000000      0.000000
25%     2005.000000      8.260000
50%     2007.000000     20.240000
75%     2009.000000     40.490000
max     2009.000000    100.000000

In [57]:
# descriptive stats
internet_activities2.describe()

TIME_PERIOD      OBS_VALUE
count  153420.000000  151587.000000
mean     2014.934207      36.421048
std         2.871363      27.998105
min      2010.000000       0.000000
25%      2013.000000      11.770000
50%      2015.000000      30.060000
75%      2017.000000      57.730000
max      2019.000000     100.000000

In [58]:
# descriptive stats
internet_activities3.describe()

TIME_PERIOD      OBS_VALUE
count  102937.000000  100156.000000
mean     2022.025365      49.301412
std         1.440421      31.028824
min      2020.000000       0.000000
25%      2021.000000      20.130000
50%      2022.000000      48.500000
75%      2023.000000      78.800000
max      2024.000000     100.000000

The three dataframes together have information from 2003 to 2024.
The values are correct, as the minimun is 0 and the maximum 100 corresponding to percentages. 
There are missing values, that would be addressed. 

## 03.3 Dropping Columns 

In [59]:
internet_activities['DATAFLOW'].value_counts(dropna = False)

DATAFLOW
ESTAT:ISOC_CI_AC_I(1.0)    56474
Name: count, dtype: int64

In [60]:
internet_activities2['DATAFLOW'].value_counts(dropna = False)

DATAFLOW
ESTAT:ISOC_CI_AC_I(1.0)    153420
Name: count, dtype: int64

In [61]:
internet_activities3['DATAFLOW'].value_counts(dropna = False)

DATAFLOW
ESTAT:ISOC_CI_AC_I(1.0)    102937
Name: count, dtype: int64

In [62]:
internet_activities['LAST UPDATE'].value_counts(dropna = False)

LAST UPDATE
24/06/25 11:00:00    56474
Name: count, dtype: int64

In [63]:
internet_activities2['LAST UPDATE'].value_counts(dropna = False)

LAST UPDATE
24/06/25 11:00:00    153420
Name: count, dtype: int64

In [64]:
internet_activities3['LAST UPDATE'].value_counts(dropna = False)

LAST UPDATE
24/06/25 11:00:00    102937
Name: count, dtype: int64

In [65]:
internet_activities['freq'].value_counts(dropna = False)

freq
Annual    56474
Name: count, dtype: int64

In [66]:
internet_activities2['freq'].value_counts(dropna = False)

freq
Annual    153420
Name: count, dtype: int64

In [67]:
internet_activities3['freq'].value_counts(dropna = False)

freq
Annual    102937
Name: count, dtype: int64

In [68]:
internet_activities['unit'].value_counts(dropna = False)

unit
Percentage of individuals    56474
Name: count, dtype: int64

In [69]:
internet_activities2['unit'].value_counts(dropna = False)

unit
Percentage of individuals    153420
Name: count, dtype: int64

In [70]:
internet_activities3['unit'].value_counts(dropna = False)

unit
Percentage of individuals    102937
Name: count, dtype: int64

In [71]:
internet_activities['OBS_FLAG'].value_counts(dropna = False)

OBS_FLAG
NaN    53329
u       3145
Name: count, dtype: int64

In [72]:
internet_activities2['OBS_FLAG'].value_counts(dropna = False)

OBS_FLAG
NaN    145724
b        3885
u        3730
bu         81
Name: count, dtype: int64

In [73]:
internet_activities3['OBS_FLAG'].value_counts(dropna = False)

OBS_FLAG
NaN    92553
u       6657
b       2376
e       1160
bu       191
Name: count, dtype: int64

This column would not be dropped, as it contains information about the values presented on those rows (flag letters). 

In [74]:
internet_activities['CONF_STATUS'].value_counts(dropna = False)

CONF_STATUS
NaN    56471
C          3
Name: count, dtype: int64

In [75]:
df_c = internet_activities[internet_activities['CONF_STATUS'] == 'C']
df_c

DATAFLOW        LAST UPDATE    freq  \
30632  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
30634  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
38513  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   

                                                indic_is  \
30632  Internet use: job search or sending an applica...   
30634  Internet use: job search or sending an applica...   
38513  Internet use: doing an online course (of any s...   

                            unit                           ind_type      geo  \
30632  Percentage of individuals  Individuals, 75 years old or more  Germany   
30634  Percentage of individuals  Individuals, 75 years old or more  Germany   
38513  Percentage of individuals  Individuals, 75 years old or more  Germany   

       TIME_PERIOD  OBS_VALUE OBS_FLAG CONF_STATUS  
30632         2006        NaN      NaN           C  
30634         2008        NaN      NaN           C  
38513         2007        NaN      NaN           C

In [76]:
internet_activities2['CONF_STATUS'].value_counts(dropna = False)

CONF_STATUS
NaN    153419
C           1
Name: count, dtype: int64

In [78]:
df_c2 = internet_activities2[internet_activities2['CONF_STATUS'] == 'C']
df_c2

DATAFLOW        LAST UPDATE    freq  \
90147  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   

                                                indic_is  \
90147  Internet use: doing an online course (of any s...   

                            unit                           ind_type      geo  \
90147  Percentage of individuals  Individuals, 75 years old or more  Germany   

       TIME_PERIOD  OBS_VALUE OBS_FLAG CONF_STATUS  
90147         2010        NaN      NaN           C

In [79]:
internet_activities3['CONF_STATUS'].value_counts(dropna = False)

CONF_STATUS
NaN    102892
C          45
Name: count, dtype: int64

In [ ]:
# Checking for more details on CONF_STATUS variable 
internet_activities[internet_activities['CONF_STATUS'].notnull()]

In [80]:
df_c3 = internet_activities3[internet_activities3['CONF_STATUS'] == 'C']
df_c3

DATAFLOW        LAST UPDATE    freq  \
87430  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
87581  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
87739  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
87921  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
88103  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
88285  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
88467  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
88649  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
88831  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
89013  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
89195  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
89493  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
89675  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
89857  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
90039  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
90217  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
90378  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
90541  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
90708  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
90889  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
91070  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
91242  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
91420  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
91602  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
91784  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
91966  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
92148  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
92330  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
92512  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
92694  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
92876  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
93057  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
93231  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
93406  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
93588  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
93770  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
93952  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
94134  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
94316  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
94498  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
94680  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
94862  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
95044  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
95226  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   
95408  ESTAT:ISOC_CI_AC_I(1.0)  24/06/25 11:00:00  Annual   

                                                indic_is  \
87430  Internet use: participating in social networks...   
87581  Internet use: participating in social networks...   
87739  Internet use: participating in social networks...   
87921  Internet use: participating in social networks...   
88103  Internet use: participating in social networks...   
88285  Internet use: participating in social networks...   
88467  Internet use: participating in social networks...   
88649  Internet use: participating in social networks...   
88831  Internet use: participating in social networks...   
89013  Internet use: participating in social networks...   
89195  Internet use: participating in social networks...   
89493  Internet use: participating in social networks...   
89675  Internet use: participating in social networks...   
89857  Internet use: participating in social networks...   
90039  Internet use: participating in social networks...   
90217  Internet use: participating in social networks...   
90378  Internet use: participating in social networks...   
90541  Internet use: participating in social networks...   
90708  Internet use: participating in social networks...   
90889  Internet

The rows in the three dataframes will be deleted as it contains no relevant information and is a very little percentage of the total data. 

In [81]:
# Deleting rows in dataframe 1
internet_activities = internet_activities[internet_activities['CONF_STATUS'] != 'C']

In [82]:
# Deleting rows in dataframe 2 
internet_activities2 = internet_activities2[internet_activities2['CONF_STATUS'] != 'C']

In [83]:
# Deleting rows in dataframe 3
internet_activities3 = internet_activities3[internet_activities3['CONF_STATUS'] != 'C']

In [85]:
# Checking results
internet_activities['CONF_STATUS'].value_counts(dropna=False)

CONF_STATUS
NaN    56471
Name: count, dtype: int64

In [86]:
# Checking results
internet_activities2['CONF_STATUS'].value_counts(dropna=False)

CONF_STATUS
NaN    153419
Name: count, dtype: int64

In [87]:
# Checking results
internet_activities3['CONF_STATUS'].value_counts(dropna=False)

CONF_STATUS
NaN    102892
Name: count, dtype: int64

In [88]:
# Drop unnecessary columns: DATAFLOW / LAST UPDATE / freq / unit / CONF_STATUS
internet_activities = internet_activities.drop(columns = ['DATAFLOW', 'LAST UPDATE', 'freq', 'unit', 'CONF_STATUS'])

In [89]:
# Drop unnecessary columns: DATAFLOW / LAST UPDATE / freq / unit / CONF_STATUS
internet_activities2 = internet_activities2.drop(columns = ['DATAFLOW', 'LAST UPDATE', 'freq', 'unit', 'CONF_STATUS'])

In [90]:
# Drop unnecessary columns: DATAFLOW / LAST UPDATE / freq / unit / CONF_STATUS
internet_activities3 = internet_activities3.drop(columns = ['DATAFLOW', 'LAST UPDATE', 'freq', 'unit', 'CONF_STATUS'])

In [91]:
# checking dataframe
internet_activities.head()

indic_is                           ind_type      geo  \
0  Internet use: Internet banking  Females with low formal education  Austria   
1  Internet use: Internet banking  Females with low formal education  Austria   
2  Internet use: Internet banking  Females with low formal education  Austria   
3  Internet use: Internet banking  Females with low formal education  Austria   
4  Internet use: Internet banking  Females with low formal education  Austria   

   TIME_PERIOD  OBS_VALUE OBS_FLAG  
0         2004       3.94      NaN  
1         2005       4.71      NaN  
2         2006       5.29      NaN  
3         2007        NaN        u  
4         2008       7.94      NaN

In [92]:
# checking dataframe
internet_activities2.head()

indic_is                     ind_type      geo  \
0  Internet use: Internet banking  Employees working full time  Albania   
1  Internet use: Internet banking  Employees working full time  Albania   
2  Internet use: Internet banking  Employees working full time  Austria   
3  Internet use: Internet banking  Employees working full time  Austria   
4  Internet use: Internet banking  Employees working full time  Austria   

   TIME_PERIOD  OBS_VALUE OBS_FLAG  
0         2018       2.83      NaN  
1         2019       2.72      NaN  
2         2017      69.41      NaN  
3         2018      69.62      NaN  
4         2019      73.70      NaN

In [93]:
# checking dataframe
internet_activities3.head()

indic_is  \
0  Internet use: accessing personal health record...   
1  Internet use: accessing personal health record...   
2  Internet use: accessing personal health record...   
3  Internet use: accessing personal health record...   
4  Internet use: accessing personal health record...   

                                            ind_type       geo  TIME_PERIOD  \
0  Persons aged 16-74 working in agriculture, for...   Austria         2022   
1  Persons aged 16-74 working in agriculture, for...   Austria         2024   
2  Persons aged 16-74 working in agriculture, for...   Belgium         2022   
3  Persons aged 16-74 working in agriculture, for...   Belgium         2024   
4  Persons aged 16-74 working in agriculture, for...  Bulgaria         2022   

   OBS_VALUE OBS_FLAG  
0      12.47      NaN  
1       9.03      NaN  
2      14.77        u  
3      46.79        u  
4      10.92      NaN

## 03.4 Renaming Columns

In [94]:
# Renaming columns for better understanding and same format
internet_activities.rename(columns = {
    'indic_is' : 'internet_activities',
    'ind_type' : 'demographic_group',
    'geo' : 'country',
    'TIME_PERIOD' : 'year',
    'OBS_VALUE' : 'internet_activities_rate',
    'OBS_FLAG' : 'activities_data_flag'
}, inplace = True)

In [95]:
internet_activities.head()

internet_activities                  demographic_group  country  \
0  Internet use: Internet banking  Females with low formal education  Austria   
1  Internet use: Internet banking  Females with low formal education  Austria   
2  Internet use: Internet banking  Females with low formal education  Austria   
3  Internet use: Internet banking  Females with low formal education  Austria   
4  Internet use: Internet banking  Females with low formal education  Austria   

   year  internet_activities_rate activities_data_flag  
0  2004                      3.94                  NaN  
1  2005                      4.71                  NaN  
2  2006                      5.29                  NaN  
3  2007                       NaN                    u  
4  2008                      7.94                  NaN

In [96]:
# Renaming columns for better understanding and same format
internet_activities2.rename(columns = {
    'indic_is' : 'internet_activities',
    'ind_type' : 'demographic_group',
    'geo' : 'country',
    'TIME_PERIOD' : 'year',
    'OBS_VALUE' : 'internet_activities_rate',
    'OBS_FLAG' : 'activities_data_flag'
}, inplace = True)

In [97]:
internet_activities2.head()

internet_activities            demographic_group  country  year  \
0  Internet use: Internet banking  Employees working full time  Albania  2018   
1  Internet use: Internet banking  Employees working full time  Albania  2019   
2  Internet use: Internet banking  Employees working full time  Austria  2017   
3  Internet use: Internet banking  Employees working full time  Austria  2018   
4  Internet use: Internet banking  Employees working full time  Austria  2019   

   internet_activities_rate activities_data_flag  
0                      2.83                  NaN  
1                      2.72                  NaN  
2                     69.41                  NaN  
3                     69.62                  NaN  
4                     73.70                  NaN

In [98]:
# Renaming columns for better understanding and same format
internet_activities3.rename(columns = {
    'indic_is' : 'internet_activities',
    'ind_type' : 'demographic_group',
    'geo' : 'country',
    'TIME_PERIOD' : 'year',
    'OBS_VALUE' : 'internet_activities_rate',
    'OBS_FLAG' : 'activities_data_flag'
}, inplace = True)

In [99]:
internet_activities3.head()

internet_activities  \
0  Internet use: accessing personal health record...   
1  Internet use: accessing personal health record...   
2  Internet use: accessing personal health record...   
3  Internet use: accessing personal health record...   
4  Internet use: accessing personal health record...   

                                   demographic_group   country  year  \
0  Persons aged 16-74 working in agriculture, for...   Austria  2022   
1  Persons aged 16-74 working in agriculture, for...   Austria  2024   
2  Persons aged 16-74 working in agriculture, for...   Belgium  2022   
3  Persons aged 16-74 working in agriculture, for...   Belgium  2024   
4  Persons aged 16-74 working in agriculture, for...  Bulgaria  2022   

   internet_activities_rate activities_data_flag  
0                     12.47                  NaN  
1                      9.03                  NaN  
2                     14.77                    u  
3                     46.79                    u  
4                     10.92                  NaN

## 03.5 Removing Prefixes

In [100]:
# Checking for variables in internet_activitiies to see if all have the same prefix - Internet use - 
internet_activities['internet_activities'].value_counts(dropna = False)

internet_activities
Internet use: finding information about goods and services                   5166
Internet use: sending/receiving e-mails                                      5165
Internet use: Internet banking                                               5133
Internet use: reading/downloading online newspapers/news                     5086
Internet use: travel and accommodation services                              5085
Internet use: listening to web radios and/or watching web TV                 5041
Internet use: job search or sending an application                           4861
Internet use: selling goods or services                                      4807
Internet use: playing/downloading games, images, films or music              4223
Internet use: downloading software                                           4217
Internet use: doing an online course (of any subject)                        2782
Internet use: telephoning or video calls                                     1

In [101]:
# Checking for variables in internet_activitiies to see if all have the same prefix - Internet use - 
internet_activities2['internet_activities'].value_counts(dropna = False)

internet_activities
Internet use: Internet banking                                                                                                                   12784
Internet use: telephoning or video calls                                                                                                         12784
Internet use: finding information about goods and services                                                                                       12781
Internet use: selling goods or services                                                                                                          12732
Internet use: sending/receiving e-mails                                                                                                          11867
Internet use: participating in social networks (creating user profile, posting messages or other contributions to facebook, twitter, etc.)       10812
Internet use: travel and accommodation services                           

In [102]:
# Checking for variables in internet_activitiies to see if all have the same prefix - Internet use - 
internet_activities3['internet_activities'].value_counts(dropna = False)

internet_activities
Internet use: Internet banking                                                                                                                10003
Internet use: sending/receiving e-mails                                                                                                       10003
Internet use: finding information about goods and services                                                                                    10003
Internet use: doing an online course (of any subject)                                                                                         10003
Internet use: telephoning or video calls                                                                                                      10003
Internet use: selling goods or services                                                                                                       10003
Internet use: participating in social networks (creating user profile, posting messages or o

In [103]:
# Removing prefix 
internet_activities['internet_activities'] = internet_activities['internet_activities'].str.replace('Internet use: ', '')

In [104]:
internet_activities['internet_activities'].value_counts(dropna = False)

internet_activities
finding information about goods and services                   5166
sending/receiving e-mails                                      5165
Internet banking                                               5133
reading/downloading online newspapers/news                     5086
travel and accommodation services                              5085
listening to web radios and/or watching web TV                 5041
job search or sending an application                           4861
selling goods or services                                      4807
playing/downloading games, images, films or music              4223
downloading software                                           4217
doing an online course (of any subject)                        2782
telephoning or video calls                                     1938
uploading self-created content to any website to be shared     1938
posting messages to social media sites or instant messaging    1029
Name: count, dtype: int64

In [105]:
# Removing prefix in second dataframe
internet_activities2['internet_activities'] = internet_activities2['internet_activities'].str.replace('Internet use: ', '')

In [106]:
internet_activities2['internet_activities'].value_counts(dropna = False)

internet_activities
Internet banking                                                                                                                   12784
telephoning or video calls                                                                                                         12784
finding information about goods and services                                                                                       12781
selling goods or services                                                                                                          12732
sending/receiving e-mails                                                                                                          11867
participating in social networks (creating user profile, posting messages or other contributions to facebook, twitter, etc.)       10812
travel and accommodation services                                                                                                   9033
uploading self-create

In [107]:
# Removing prefix in second dataframe
internet_activities3['internet_activities'] = internet_activities3['internet_activities'].str.replace('Internet use: ', '')

In [108]:
internet_activities3['internet_activities'].value_counts(dropna = False)

internet_activities
Internet banking                                                                                                                10003
sending/receiving e-mails                                                                                                       10003
finding information about goods and services                                                                                    10003
doing an online course (of any subject)                                                                                         10003
telephoning or video calls                                                                                                      10003
selling goods or services                                                                                                       10003
participating in social networks (creating user profile, posting messages or other contributions to facebook, twitter, etc.)     9958
civic or political participation          

## 03.6 Mixed Datatypes

In [109]:
# function to check for mixed-type columns
for col in internet_activities.columns.tolist():
    weird = (internet_activities[[col]].applymap(type) != internet_activities[[col]].iloc[0].apply(type)).any(axis=1)
    if len (internet_activities[weird]) > 0:
        print(col)

activities_data_flag


C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\1415708128.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities[[col]].applymap(type) != internet_activities[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\1415708128.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities[[col]].applymap(type) != internet_activities[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\1415708128.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities[[col]].applymap(type) != internet_activities[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\1415708128.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities[[col]].applymap(type) != 

In [110]:
# function to check for mixed-type columns
for col in internet_activities2.columns.tolist():
    weird = (internet_activities2[[col]].applymap(type) != internet_activities2[[col]].iloc[0].apply(type)).any(axis=1)
    if len (internet_activities2[weird]) > 0:
        print(col)

C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\3158159759.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities2[[col]].applymap(type) != internet_activities2[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\3158159759.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities2[[col]].applymap(type) != internet_activities2[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\3158159759.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities2[[col]].applymap(type) != internet_activities2[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\3158159759.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities2[[col]].applymap(ty

activities_data_flag


C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\3158159759.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities2[[col]].applymap(type) != internet_activities2[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\3158159759.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities2[[col]].applymap(type) != internet_activities2[[col]].iloc[0].apply(type)).any(axis=1)


In [111]:
# function to check for mixed-type columns
for col in internet_activities3.columns.tolist():
    weird = (internet_activities3[[col]].applymap(type) != internet_activities3[[col]].iloc[0].apply(type)).any(axis=1)
    if len (internet_activities3[weird]) > 0:
        print(col)

C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\959092838.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities3[[col]].applymap(type) != internet_activities3[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\959092838.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities3[[col]].applymap(type) != internet_activities3[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\959092838.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities3[[col]].applymap(type) != internet_activities3[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\959092838.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (internet_activities3[[col]].applymap(type) 

activities_data_flag


In [112]:
# Changing data type for activities_data_flag variable
internet_activities['activities_data_flag'] = internet_activities['activities_data_flag'].astype('category')

In [113]:
# Changing data type for activities_data_flag variable
internet_activities2['activities_data_flag'] = internet_activities2['activities_data_flag'].astype('category')

In [114]:
# Changing data type for activities_data_flag variable
internet_activities3['activities_data_flag'] = internet_activities3['activities_data_flag'].astype('category')

In [116]:
internet_activities.dtypes

internet_activities           object
demographic_group             object
country                       object
year                           int64
internet_activities_rate     float64
activities_data_flag        category
dtype: object

In [117]:
internet_activities2.dtypes

internet_activities           object
demographic_group             object
country                       object
year                           int64
internet_activities_rate     float64
activities_data_flag        category
dtype: object

In [118]:
internet_activities3.dtypes

internet_activities           object
demographic_group             object
country                       object
year                           int64
internet_activities_rate     float64
activities_data_flag        category
dtype: object

## 03.7 Missing Values

### 03.7.1 Dataframe internet activities

In [119]:
# Finding missing values in internet_activities
internet_activities.isnull().sum()

internet_activities             0
demographic_group               0
country                         0
year                            0
internet_activities_rate     3094
activities_data_flag        53326
dtype: int64

In [121]:
internet_activities.shape

(56471, 6)

In [120]:
# Creating subset for NaN values 
df_nan = internet_activities[internet_activities['internet_activities_rate'].isnull() == True]
df_nan

internet_activities  \
3                                    Internet banking   
12                                   Internet banking   
13                                   Internet banking   
14                                   Internet banking   
15                                   Internet banking   
...                                               ...   
56469  listening to web radios and/or watching web TV   
56470  listening to web radios and/or watching web TV   
56471  listening to web radios and/or watching web TV   
56472  listening to web radios and/or watching web TV   
56473  listening to web radios and/or watching web TV   

                       demographic_group         country  year  \
3      Females with low formal education         Austria  2007   
12     Females with low formal education        Bulgaria  2006   
13     Females with low formal education        Bulgaria  2007   
14     Females with low formal education        Bulgaria  2008   
15     Females with low formal education        Bulgaria  2009   
...                                  ...             ...   ...   
56469  Individuals, 75 years old or more  United Kingdom  2005   
56470  Individuals, 75 years old or more  United Kingdom  2006   
56471  Individuals, 75 years old or more  United Kingdom  2007   
56472  Individuals, 75 years old or more  United Kingdom  2008   
56473  Individuals, 75 years old or more  United Kingdom  2009   

       internet_activities_rate activities_data_flag  
3                           NaN                    u  
12                          NaN                    u  
13                          NaN                    u  
14                          NaN                    u  
15                          NaN                    u  
...                         ...                  ...  
56469                       NaN                    u  
56470                       NaN                    u  
56471                       NaN                    u  
56472                       NaN                    u  
56473                       NaN                    u  

[3094 rows x 6 columns]

In [122]:
# Exploring more on the NaN values
df_nan['activities_data_flag'].value_counts()

activities_data_flag
u    3094
Name: count, dtype: int64

In [123]:
# Exploring more on the NaN values
df_nan['country'].value_counts()

country
Germany            442
North Macedonia    397
United Kingdom     347
Portugal           278
Austria            271
Slovenia           244
Italy              207
Estonia            138
Romania            116
Poland             109
Bulgaria           103
Ireland             65
Iceland             59
Lithuania           57
Czechia             52
Malta               47
Denmark             45
Hungary             31
France              27
Slovakia            24
Greece              23
Latvia              11
Finland              1
Name: count, dtype: int64

In [124]:
# Exploring more on the NaN values
df_nan['year'].value_counts()

year
2005    647
2009    564
2008    522
2007    446
2006    418
2004    393
2003    104
Name: count, dtype: int64

The missing values will be imputed with the mean for each year and country. 

In [127]:
# Creating a copy of dataframe to impute values
internet_activities_clean = internet_activities.copy()

In [129]:
# Imputing values with mean of country and year
internet_activities_clean['internet_activities_rate'] = (internet_activities.groupby('country')['internet_activities_rate'].transform(lambda x: x.fillna(x.mean())))

In [131]:
internet_activities_clean.isnull().sum()

internet_activities             0
demographic_group               0
country                         0
year                            0
internet_activities_rate        0
activities_data_flag        53326
dtype: int64

### 03.7.2 Dataframe internet activities 2

In [133]:
# Finding missing values in internet_activities2
internet_activities2.isnull().sum()

internet_activities              0
demographic_group                0
country                          0
year                             0
internet_activities_rate      1832
activities_data_flag        145723
dtype: int64

In [134]:
# Creating subset for NaN values 
df_nan2 = internet_activities2[internet_activities2['internet_activities_rate'].isnull() == True]
df_nan2

internet_activities  \
164                                      Internet banking   
248                                      Internet banking   
382                                      Internet banking   
385                                      Internet banking   
581                                      Internet banking   
...                                                   ...   
152169  consulting wikis (to obtain knowledge on any s...   
152449  consulting wikis (to obtain knowledge on any s...   
152951  consulting wikis (to obtain knowledge on any s...   
153000  consulting wikis (to obtain knowledge on any s...   
153001  consulting wikis (to obtain knowledge on any s...   

                                   demographic_group          country  year  \
164                      Employees working part time           Serbia  2019   
248                Females with low formal education          Denmark  2015   
382                Females with low formal education       Montenegro  2017   
385                Females with low formal education  North Macedonia  2010   
581             Females with medium formal education          Denmark  2015   
...                                              ...              ...   ...   
152169              Males with high formal education          Denmark  2015   
152449                                    Unemployed          Denmark  2011   
152951  Individuals aged 25 to 64 who are unemployed          Denmark  2011   
153000  Individuals aged 25 to 64 who are unemployed           Norway  2011   
153001  Individuals aged 25 to 64 who are unemployed           Norway  2013   

        internet_activities_rate activities_data_flag  
164                          NaN                    u  
248                          NaN                    u  
382                          NaN                    u  
385                          NaN                    u  
581                          NaN                    u  
...                          ...                  ...  
152169                       NaN                    u  
152449                       NaN                    u  
152951                       NaN                    u  
153000                       NaN                    u  
153001                       NaN                    u  

[1832 rows x 6 columns]

In [136]:
## Exploring more on the NaN values
df_nan2['activities_data_flag'].value_counts()

activities_data_flag
u     1832
b        0
bu       0
Name: count, dtype: int64

In [137]:
# Exploring more on the NaN values
df_nan2['country'].value_counts()

country
Italy              239
Denmark            219
North Macedonia    203
Iceland            139
Malta              136
France             102
Ireland             99
Norway              93
United Kingdom      86
Montenegro          75
Sweden              64
Luxembourg          54
Portugal            51
Austria             48
Belgium             32
Lithuania           27
Greece              26
Germany             24
Czechia             24
Bulgaria            22
Cyprus              15
Slovenia            14
Albania             11
Serbia               9
Kosovo*              8
Romania              6
Poland               6
Name: count, dtype: int64

In [138]:
# Exploring more on the NaN values
df_nan2['year'].value_counts()

year
2010    482
2015    315
2017    193
2019    168
2014    146
2012    117
2018    116
2011    106
2016     98
2013     91
Name: count, dtype: int64

The missing values will be imputed with the mean for each year and country. 

In [139]:
# Creating a copy of dataframe to impute values
internet_activities_clean2 = internet_activities2.copy()

In [140]:
# Imputing values with mean of country and year
internet_activities_clean2['internet_activities_rate'] = (internet_activities2.groupby('country')['internet_activities_rate'].transform(lambda x: x.fillna(x.mean())))

In [141]:
# Checking results
internet_activities_clean2.isnull().sum()

internet_activities              0
demographic_group                0
country                          0
year                             0
internet_activities_rate         0
activities_data_flag        145723
dtype: int64

### 03.7.3 Dataframe interent activities 3

In [142]:
# Finding missing values in internet_activities3
internet_activities3.isnull().sum()

internet_activities             0
demographic_group               0
country                         0
year                            0
internet_activities_rate     2736
activities_data_flag        92508
dtype: int64

In [143]:
# Creating subset for NaN values 
df_nan3 = internet_activities3[internet_activities3['internet_activities_rate'].isnull() == True]
df_nan3

internet_activities  \
11               accessing personal health records online   
44               accessing personal health records online   
45               accessing personal health records online   
46               accessing personal health records online   
47               accessing personal health records online   
...                                                   ...   
102107  sharing or publishing self-created videos, pho...   
102139  sharing or publishing self-created videos, pho...   
102284  sharing or publishing self-created videos, pho...   
102506  sharing or publishing self-created videos, pho...   
102580  sharing or publishing self-created videos, pho...   

                                        demographic_group          country  \
11      Persons aged 16-74 working in agriculture, for...          Denmark   
44      Persons aged 16-74 working in mining or quarrying          Austria   
45      Persons aged 16-74 working in mining or quarrying          Austria   
46      Persons aged 16-74 working in mining or quarrying          Belgium   
47      Persons aged 16-74 working in mining or quarrying          Belgium   
...                                                   ...              ...   
102107                                  ICT professionals  North Macedonia   
102139                              Non ICT professionals            Italy   
102284                          Males, 16 to 24 years old          Ireland   
102506                                           Students          Ireland   
102580                    Individuals, 16 to 24 years old          Ireland   

        year  internet_activities_rate activities_data_flag  
11      2024                       NaN                    u  
44      2022                       NaN                    u  
45      2024                       NaN                    u  
46      2022                       NaN                    u  
47      2024                       NaN                    u  
...      ...                       ...                  ...  
102107  2020                       NaN                    u  
102139  2020                       NaN                    u  
102284  2020                       NaN                    u  
102506  2020                       NaN                    u  
102580  2020                       NaN                    u  

[2736 rows x 6 columns]

In [145]:
## Exploring more on the NaN values
df_nan3['activities_data_flag'].value_counts()

activities_data_flag
u     2634
bu     102
b        0
e        0
Name: count, dtype: int64

In [152]:
## Exploring more on the NaN values
df_nan3['country'].value_counts()

country
Ireland            291
Slovakia           194
Luxembourg         180
Denmark            173
Greece             162
Croatia            160
Cyprus             160
Slovenia           158
Norway             149
Netherlands        120
Belgium            104
Sweden             102
Austria            100
Estonia             80
United Kingdom      73
Portugal            60
Lithuania           58
Iceland             54
Hungary             51
Serbia              50
Czechia             48
Italy               42
Albania             33
Germany             31
Latvia              22
North Macedonia     21
Romania             20
Bulgaria            20
Montenegro          11
Spain                9
Name: count, dtype: int64

In [147]:
## Exploring more on the NaN values
df_nan3['year'].value_counts()

year
2022    714
2021    639
2024    636
2023    576
2020    171
Name: count, dtype: int64

The missing values will be imputed with the mean for each year and country.

In [148]:
# Creating a copy of dataframe to impute values
internet_activities_clean3 = internet_activities3.copy()

In [149]:
# Imputing values with mean of country and year
internet_activities_clean3['internet_activities_rate'] = (internet_activities3.groupby('country')['internet_activities_rate'].transform(lambda x: x.fillna(x.mean())))

In [150]:
# Checking results
internet_activities_clean3.isnull().sum()

internet_activities             0
demographic_group               0
country                         0
year                            0
internet_activities_rate        0
activities_data_flag        92508
dtype: int64

## 03.8 Duplicate Values

In [156]:
# Creating a new subset 'df_dups' containing just duplicated values
activities_dups = internet_activities_clean[internet_activities_clean.duplicated()]
activities_dups

Empty DataFrame
Columns: [internet_activities, demographic_group, country, year, internet_activities_rate, activities_data_flag]
Index: []

In [157]:
# Creating a new subset 'df_dups' containing just duplicated values
activities_dups2 = internet_activities_clean2[internet_activities_clean2.duplicated()]
activities_dups2

Empty DataFrame
Columns: [internet_activities, demographic_group, country, year, internet_activities_rate, activities_data_flag]
Index: []

In [158]:
# Creating a new subset 'df_dups' containing just duplicated values
activities_dups3 = internet_activities_clean3[internet_activities_clean3.duplicated()]
activities_dups3

Empty DataFrame
Columns: [internet_activities, demographic_group, country, year, internet_activities_rate, activities_data_flag]
Index: []

There are no duplicated rows. 

## 03.9 Exporting Data

In [159]:
# Exporting first dataframe internet_use_clean
internet_activities_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'internet_activities_checked.csv'))

In [160]:
# Exporting first dataframe internet_use_clean
internet_activities_clean2.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'internet_activities2_checked.csv'))

In [161]:
# Exporting first dataframe internet_use_clean
internet_activities_clean3.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'internet_activities3_checked.csv'))

# 04. Dataset: Purchasing Power GDP

## 04.1 Import Data

In [162]:
# Import dataset (purschasing power GDP per capita)
purchasing_power = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'purchasing_power.csv'), index_col = False)

## 04.2 Data Consistency Check

In [163]:
# dimensions of dataframe
purchasing_power.shape

(809, 11)

In [164]:
# names of columns
purchasing_power.columns

Index(['DATAFLOW', 'LAST UPDATE', 'freq', 'na_item', 'ppp_cat', 'unit', 'geo',
       'TIME_PERIOD', 'OBS_VALUE', 'OBS_FLAG', 'CONF_STATUS'],
      dtype='object')

In [165]:
# first rows
purchasing_power.head()

DATAFLOW        LAST UPDATE    freq  \
0  ESTAT:SDG_10_10(1.0)  10/07/25 11:00:00  Annual   
1  ESTAT:SDG_10_10(1.0)  10/07/25 11:00:00  Annual   
2  ESTAT:SDG_10_10(1.0)  10/07/25 11:00:00  Annual   
3  ESTAT:SDG_10_10(1.0)  10/07/25 11:00:00  Annual   
4  ESTAT:SDG_10_10(1.0)  10/07/25 11:00:00  Annual   

                                          na_item                 ppp_cat  \
0  Real expenditure per capita (in PPS_EU27_2020)  Gross domestic product   
1  Real expenditure per capita (in PPS_EU27_2020)  Gross domestic product   
2  Real expenditure per capita (in PPS_EU27_2020)  Gross domestic product   
3  Real expenditure per capita (in PPS_EU27_2020)  Gross domestic product   
4  Real expenditure per capita (in PPS_EU27_2020)  Gross domestic product   

         unit      geo  TIME_PERIOD  OBS_VALUE OBS_FLAG  CONF_STATUS  
0  Percentage  Albania         2003       4200      NaN          NaN  
1  Percentage  Albania         2004       4600      NaN          NaN  
2  Percentage  Albania         2005       4900      NaN          NaN  
3  Percentage  Albania         2006       5400      NaN          NaN  
4  Percentage  Albania         2007       6000      NaN          NaN

There are columns that are unnecessary and will be deleted. 

In [166]:
# general information of dataframe
purchasing_power.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 809 entries, 0 to 808
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   DATAFLOW     809 non-null    object 
 1   LAST UPDATE  809 non-null    object 
 2   freq         809 non-null    object 
 3   na_item      809 non-null    object 
 4   ppp_cat      809 non-null    object 
 5   unit         809 non-null    object 
 6   geo          809 non-null    object 
 7   TIME_PERIOD  809 non-null    int64  
 8   OBS_VALUE    809 non-null    int64  
 9   OBS_FLAG     16 non-null     object 
 10  CONF_STATUS  0 non-null      float64
dtypes: float64(1), int64(2), object(8)
memory usage: 69.7+ KB


The variables 'CONF_STATUS' can be deleted, as it contains no information

In [167]:
# descriptive stats
purchasing_power.describe()

TIME_PERIOD     OBS_VALUE  CONF_STATUS
count   809.000000    809.000000          0.0
mean   2013.559951  26781.953028          NaN
std       6.321359  14386.929252          NaN
min    2003.000000   4200.000000          NaN
25%    2008.000000  16900.000000          NaN
50%    2014.000000  25400.000000          NaN
75%    2019.000000  33100.000000          NaN
max    2024.000000  96200.000000          NaN

## 04.3 Dropping Columns

In [168]:
# Checking columns to drop
purchasing_power['DATAFLOW'].value_counts(dropna = False)

DATAFLOW
ESTAT:SDG_10_10(1.0)    809
Name: count, dtype: int64

In [169]:
# Checking columns to drop
purchasing_power['LAST UPDATE'].value_counts(dropna = False)

LAST UPDATE
10/07/25 11:00:00    809
Name: count, dtype: int64

In [170]:
# Checking columns to drop
purchasing_power['freq'].value_counts(dropna = False)

freq
Annual    809
Name: count, dtype: int64

In [171]:
# Checking columns to drop
purchasing_power['na_item'].value_counts(dropna = False)

na_item
Real expenditure per capita (in PPS_EU27_2020)    809
Name: count, dtype: int64

In [172]:
# Checking columns to drop
purchasing_power['ppp_cat'].value_counts(dropna = False)

ppp_cat
Gross domestic product    809
Name: count, dtype: int64

In [173]:
# Checking columns to drop
purchasing_power['unit'].value_counts(dropna = False)

unit
Percentage    809
Name: count, dtype: int64

In [174]:
# Checking columns to drop
purchasing_power['OBS_FLAG'].value_counts(dropna = False)

OBS_FLAG
NaN    793
p       13
e        3
Name: count, dtype: int64

This variable will be kept, as it provides information on the values provided. 

In [175]:
# Checking columns to drop
purchasing_power['CONF_STATUS'].value_counts(dropna = False)

CONF_STATUS
NaN    809
Name: count, dtype: int64

In [176]:
# Drop unnecessary columns: DATAFLOW / LAST UPDATE / freq / unit / CONF_STATUS
purchasing_power = purchasing_power.drop(columns = ['DATAFLOW', 'LAST UPDATE', 'freq', 'na_item', 'ppp_cat', 'unit', 'CONF_STATUS'])

In [177]:
# Checking dataframe
purchasing_power.head()

geo  TIME_PERIOD  OBS_VALUE OBS_FLAG
0  Albania         2003       4200      NaN
1  Albania         2004       4600      NaN
2  Albania         2005       4900      NaN
3  Albania         2006       5400      NaN
4  Albania         2007       6000      NaN

## 04.4 Renaming Columns

In [178]:
# Renaming columns for better understanding and same format
purchasing_power.rename(columns = {
    'geo' : 'country',
    'TIME_PERIOD' : 'year',
    'OBS_VALUE' : 'purchasing_power_gdp',
    'OBS_FLAG' : 'pp_gdp_data_flag'
}, inplace = True)

In [180]:
purchasing_power.head()

country  year  purchasing_power_gdp pp_gdp_data_flag
0  Albania  2003                  4200              NaN
1  Albania  2004                  4600              NaN
2  Albania  2005                  4900              NaN
3  Albania  2006                  5400              NaN
4  Albania  2007                  6000              NaN

## 04.5 Mixed Data Types

In [181]:
# function to check for mixed-type columns
for col in purchasing_power.columns.tolist():
    weird = (purchasing_power[[col]].applymap(type) != purchasing_power[[col]].iloc[0].apply(type)).any(axis=1)
    if len (purchasing_power[weird]) > 0:
        print(col)

pp_gdp_data_flag


C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\1689887623.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (purchasing_power[[col]].applymap(type) != purchasing_power[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\1689887623.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (purchasing_power[[col]].applymap(type) != purchasing_power[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\1689887623.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (purchasing_power[[col]].applymap(type) != purchasing_power[[col]].iloc[0].apply(type)).any(axis=1)
C:\Users\adrig\AppData\Local\Temp\ipykernel_35692\1689887623.py:3: FutureWarning: DataFrame.applymap has been deprecated. Use DataFrame.map instead.
  weird = (purchasing_power[[col]].applymap(type) != purchasing_power[[col

In [182]:
# Changing data type for activities_data_flag variable
purchasing_power['pp_gdp_data_flag'] = purchasing_power['pp_gdp_data_flag'].astype('category')

In [183]:
purchasing_power.dtypes

country                   object
year                       int64
purchasing_power_gdp       int64
pp_gdp_data_flag        category
dtype: object

## 04.6 Missing Data

In [184]:
#looking for missing data 
purchasing_power.isnull().sum()

country                   0
year                      0
purchasing_power_gdp      0
pp_gdp_data_flag        793
dtype: int64

There is no missing data (the missing data in pp_gdp_data_flag refers to the values that do not have any observation flag). 

## 04.7 Duplicate Values

In [185]:
# Creating a subset for the duplicate values
pp_dups = purchasing_power[purchasing_power.duplicated()]

In [186]:
pp_dups

Empty DataFrame
Columns: [country, year, purchasing_power_gdp, pp_gdp_data_flag]
Index: []

There are no duplicated values.

## 04.8 Exporting Data

In [187]:
# Exporting first dataframe internet_use_clean
purchasing_power.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'purchasing_power_gdp_checked.csv'))